Include prebuilt UI assets (or tarball) in release artifacts

### Proposal

Since 0.32.0 (as per #5113), compiled UI assets are no longer included in the repository. This means downstream builds now need to build the UI before building the Go binary.

As the downstream maintainer for Fedora & EPEL, this introduces a significant complication in the packaging process. Our previous spec file worked without modification, but now fails with:

    ui/web.go:31:12: pattern app/dist: no matching files found

This change requires adding a full frontend (Elm + tooling) build pipeline into the RPM build, which is considerably more complex and less reproducible in our environment. In addition, such toolchains are often problematic in distribution build systems due to policy and network restrictions.

For comparison, Prometheus ships prebuilt UI assets as part of its release process (`prometheus-web-ui-$version.tar.gz`). This allows downstream packagers to:
- Avoid rebuilding frontend assets
- Keep builds simpler and more reproducible
- Only declare bundled JS dependencies instead of bootstrapping a full toolchain

Would it be possible to publish a tarball with prebuilt UI assets as part of Alertmanager releases (or otherwise include them in release artifacts)?

This would greatly simplify downstream packaging while preserving the current development workflow.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -22,15 +22,29 @@ jobs:
       - run: make assets
       - run: make apiv2
       - run: git diff --exit-code
+      - run: make assets-tarball
+      - uses: prometheus/promci-artifacts/save@f9a587dbc0b2c78a0c54f8ad1cde71ea29a4b76f # v0.1.0
+        with:
+          directory: .tarballs
+      - uses: actions/upload-artifact@bbbca2ddaa5d8feaa63e36b76fdaad77386f024f # v4.6.2
+        with:
+          name: ui-dist
+          path: ui/app/dist
+          include-hidden-files: true
 
   build:
     name: Build Alertmanager for common architectures
+    needs: [test_frontend]
     runs-on: ubuntu-latest
     strategy:
       matrix:
         thread: [0, 1, 2]
     steps:
       - uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v4.0.0
+      - uses: actions/download-artifact@fa0a91b85d4f404e444e00e005971372dc801d16 # v4.1.8
+        with:
+          name: ui-dist
+          path: ui/app/dist
       - uses: prometheus/promci/build@42c3c84c865e5c1ab78543b929f7341eb2ef6123 # v0.6.1
         with:
           promu_opts: "-p linux/amd64 -p windows/amd64 -p linux/arm64 -p darwin/amd64 -p darwin/arm64 -p linux/386"
@@ -40,6 +54,7 @@ jobs:
   test:
     name: Test
     runs-on: ubuntu-latest
+    needs: [test_frontend]
     # Whenever the Go version is updated here, .promu.yml
     # should also be updated.
     container:
@@ -58,14 +73,10 @@ jobs:
       EMAIL_AUTH_CONFIG: testdata/auth.yml
     steps:
       - uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6.0.2
-      - id: cache-paths
-        run: echo "npm-cache=$(npm config get cache)" >> $GITHUB_OUTPUT
-      - uses: actions/cache@668228422ae6a00e4ad889ee87cd7109ec5666a7 # v5.0.4
+      - uses: actions/download-artifact@fa0a91b85d4f404e444e00e005971372dc801d16 # v4.1.8
         with:
-          path: ${{ steps.cache-paths.outputs.npm-cache }}
-          key: ${{ runner.os }}-node-${{ hashFiles('**/package-lock.json') }}
-          restore-keys: |
-            ${{ runner.os }}-node-
+          name: ui-dist
+          path: ui/app/dist
       - uses: prometheus/promci-setup@5af30ba8c199a91d6c04ebdc3c48e630e355f62d # v0.1.0
       - run: make
       - run: git diff --exit-code

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -18,6 +18,10 @@ jobs:
     needs: ci
     steps:
       - uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6.0.2
+      - uses: actions/download-artifact@fa0a91b85d4f404e444e00e005971372dc801d16 # v4.1.8
+        with:
+          name: ui-dist
+          path: ui/app/dist
       - uses: prometheus/promci/build@42c3c84c865e5c1ab78543b929f7341eb2ef6123 # v0.6.1
         with:
           parallelism: 12

--- a/Makefile
+++ b/Makefile
@@ -38,6 +38,11 @@ lint: ui-elm common-lint
 .PHONY: assets
 assets: $(FRONTEND_DIR)/src/Data ui-elm template/email.tmpl
 
+.PHONY: assets-tarball
+assets-tarball: ui-elm
+	mkdir -p .tarballs
+	tar czf ".tarballs/alertmanager-web-ui-$(file <VERSION).tar.gz" -C ui/app dist
+
 .PHONY: ui-elm
 ui-elm:
 	cd $(FRONTEND_DIR) && $(MAKE) build

--- a/ui/app/Makefile
+++ b/ui/app/Makefile
@@ -82,12 +82,13 @@ else
 endif
 
 .PHONY: build
-build: .build_stamp
+build: dist/.build_stamp
 
-.build_stamp: $(ELM_FILES) index.html vite.config.mjs node_modules
+dist/.build_stamp: $(ELM_FILES) index.html vite.config.mjs package-lock.json
 	@echo ">> building frontend"
+	$(MAKE) node_modules
 	$(NPM) run build
-	@touch .build_stamp
+	@touch dist/.build_stamp
 
 src/Data: ../../api/v2/openapi.yaml
 	-rm -rf src/Data src/DateTime.elm
@@ -98,8 +99,8 @@ src/Data: ../../api/v2/openapi.yaml
 	  -v ${PWD}/src:/output/src \
 	  $(OPENAPI_GEN_IMAGE) \
 	  -c 'java -jar /opt/openapi-generator/modules/openapi-generator-cli/target/openapi-generator-cli.jar generate -i /local/api/v2/openapi.yaml -g elm -o /tmp/openapi-gen && cp -r /tmp/openapi-gen/src/Data /output/src/Data && cp /tmp/openapi-gen/src/DateTime.elm /output/src/DateTime.elm'
-	make format
+	$(MAKE) format
 
 .PHONY: clean
 clean:
-	- @rm -rf .build_stamp dist elm-stuff src/Data src/DateTime.elm openapi-* elm-*
+	- @rm -rf dist node_modules elm-stuff src/Data src/DateTime.elm openapi-* elm-*

